src/Core/Algorithms/Math/ComputeTensorUncertaintyAlgorithm.cc

```diff
@@ -0,0 +1,443 @@
+  Dyadic3DTensor ret = Dyadic3DTensor(mean);
+  return ret;
+}
+
+ComputeTensorUncertaintyAlgorithmImpl::ComputeTensorUncertaintyAlgorithmImpl()
+{}
+
+void ComputeTensorUncertaintyAlgorithmImpl::computeCovarianceMatrices()
+{
+  covarianceMatrices_ = std::vector<DyadicTensor<6>>(fieldSize_);
+
+  Eigen::Matrix<double, 6, 6> covarianceMat;
+  for (size_t t = 0; t < fieldSize_; ++t)
+  {
+    covarianceMat.fill(0.0);
+    for (size_t f = 0; f < fieldCount_; ++f)
+    {
+      Eigen::Matrix<double, 6, 1> diffTensor = (tensors_[f][t] - meanTensors_[t]).mandel();
+      covarianceMat += diffTensor * diffTensor.transpose();
+    }
+    covarianceMat /= static_cast<double>(fieldCount_);
+    covarianceMatrices_[t] = DyadicTensor<6>(covarianceMat);
+  }
+}
+
+void ComputeTensorUncertaintyAlgorithmImpl::getPointsForFields(
+    FieldHandle field, std::vector<int>& indices, std::vector<Point>& points)
+{
+  // Collect indices and points from facades
+  FieldDataType fieldLocation;
+  FieldInformation finfo(field);
+  if (finfo.is_point() || finfo.is_linear())
+    fieldLocation = FieldDataType::node;
+  else if (finfo.is_line())
+    fieldLocation = FieldDataType::edge;
+  else if (finfo.is_surface())
+    fieldLocation = FieldDataType::face;
+  else
+    fieldLocation = FieldDataType::cell;
+
+  auto mesh = field->vmesh();
+  auto primaryFacade = field->mesh()->getFacade();
+  switch (fieldLocation)
+  {
+  case FieldDataType::node:
+    for (const auto& node : primaryFacade->nodes())
+    {
+      indices.push_back(node.index());
+      Point p;
+      mesh->get_center(p, node.index());
+      points.push_back(p);
+    }
+    break;
+  case FieldDataType::edge:
+    for (const auto& edge : primaryFacade->edges())
+    {
+      indices.push_back(edge.index());
+      Point p;
+      mesh->get_center(p, edge.index());
+      points.push_back(p);
+    }
+    break;
+  case FieldDataType::face:
+    for (const auto& face : primaryFacade->faces())
+    {
+      indices.push_back(face.index());
+      Point p;
+      mesh->get_center(p, face.index());
+      points.push_back(p);
+    }
+    break;
+  case FieldDataType::cell:
+    for (const auto& cell : primaryFacade->cells())
+    {
+      indices.push_back(cell.index());
+      Point p;
+      mesh->get_center(p, cell.index());
+      points.push_back(p);
+    }
+    break;
+  }
+}
+
+void ComputeTensorUncertaintyAlgorithmImpl::getPoints(const FieldList& fields)
+{
+  fieldCount_ = fields.size();
+
+  indices_ = std::vector<std::vector<int>>(fieldCount_);
+  points_ = std::vector<std::vector<Point>>(fieldCount_);
+
+  for (size_t f = 0; f < fieldCount_; ++f)
+    getPointsForFields(fields[f], indices_[f], points_[f]);
+}
+
+void ComputeTensorUncertaintyAlgorithmImpl::verifyData(const FieldList& fields)
+{
+  fieldSize_ = indices_[0].size();
+  for (size_t f = 1; f < fieldCount_; ++f)
+    if (indices_[f].size() != fieldSize_)
+      throw std::invalid_argument("All field inputs must have the same size.");
```

Comment:
Check out file `Core/Algorithms/Base/AlgorithmPreconditions.h`. This looks like a `THROW_ALGORITHM_INPUT_ERROR`

Reply:
I can't use this and I think it is because the impl class isn't an algorithm class. I'll use THROW_INVALID_ARGUMENT instead.

Reply:
There's THROW_ALGORITHM_INPUT_ERROR_SIMPLE for use outside algo classes, and THROW_ALGORITHM_INPUT_ERROR_WITH can accept a pointer to the algo class if the impl class has one saved.

Reply:
Does it matter which one I use?

Reply:
Whichever you prefer/looks "nicer" to you